app/power.c

```diff
@@ -0,0 +1,107 @@
+#include "stm32l4xx_hal.h"
+
+#include "FreeRTOS.h"
+#include "task.h"
+#include "queue.h"
+
+#include "drivers/adc.h"
+#include "drivers/gpio.h"
+#include "drivers/mcp4018t.h"
+#include "power.h"
+
+/* TODO: Make sure these values match reality */
+#define MIN_MV 1700
+#define MAX_MV 4500
+
+#define POWER_TASK_STACK_SIZE	512
+#define POWER_TASK_NAME			"Power"
+#define POWER_TASK_PRIORITY		1
+
+#define QUEUE_LENGTH			10
+#define QUEUE_ITEM_SIZE			(sizeof(uint16_t) * ADC_NUM_OF_CHANNELS)
+
+static StackType_t power_task_stack[POWER_TASK_STACK_SIZE];
+static TaskHandle_t power_task_handle;
+static StaticTask_t power_task_tcb;
+static QueueHandle_t power_queue_handle;
+static StaticQueue_t power_queue;
+static uint8_t power_queue_storage[QUEUE_LENGTH * QUEUE_ITEM_SIZE];
+
+static bool m_initialized;
+static bool m_dut_vdd_enabled;
+static uint16_t m_ldo_voltage;
+
+void power_dut_set_enabled(bool enabled)
+{
+	gpio_write(DUT_VDD_EN_GPIO_Port, DUT_VDD_EN_Pin, enabled);
+
+	m_dut_vdd_enabled = enabled;
+}
+
+err_t power_dut_ldo_set(uint32_t millivolt)
+{
+	err_t r;
+
+	if (!m_initialized)
+		return EPOWER_NO_INIT;
+
+	if (millivolt < MIN_MV || millivolt > MAX_MV)
+		return EPOWER_VOLTAGE_RANGE;
+
+	const float factor = (((float)millivolt) - MIN_MV) / (MAX_MV - MIN_MV);
+	const uint8_t value = (uint8_t) (factor * 127);
+
+	r = mcp4018t_set_value(value);
+	ERR_CHECK(r);
+
+	m_ldo_voltage = millivolt;
+
+	return r;
+}
+
+void adc_conversion_ready_handler(const uint16_t adc_values[ADC_NUM_OF_CHANNELS])
```

Comment:
Not uint16_t * ?

Reply:
Which line?

Reply:
The comments are for the line that is just above the content so #L62

Reply:
Ah. No the idea here is that by doing like this i don't need to pass a pointer + the number of values. Fewer things to do in the interrupt handler. The consumer of the API also understands how many items there are, so there can't be any confusion.

The generated assembly when doing like this or using a pointer is the same.

You can see the change in the API in the other commit in this PR. (https://github.com/xil-se/xildebug_sw/pull/24/files#diff-6f418b016d24c723ec049507e7846810R12)